homeassistant/components/compensation/sensor.py

```diff
@@ -0,0 +1,164 @@
+"""Support for compensation sensor."""
+import logging
+import re
+
+import numpy as np
+
+from homeassistant.const import (
+    ATTR_ENTITY_ID,
+    ATTR_UNIT_OF_MEASUREMENT,
+    CONF_ATTRIBUTE,
+    CONF_ENTITY_ID,
+    CONF_NAME,
+    CONF_UNIT_OF_MEASUREMENT,
+    STATE_UNKNOWN,
+)
+from homeassistant.core import callback
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.event import async_track_state_change_event
+
+from .const import (
+    CONF_COMPENSATION,
+    CONF_DATAPOINTS,
+    CONF_DEGREE,
+    CONF_PRECISION,
+    DATA_COMPENSATION,
+    MATCH_DATAPOINT,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+ATTR_ATTRIBUTE = "attribute"
+ATTR_COEFFICIENTS = "coefficients"
+
+
+async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
+    """Set up the Compensation sensor."""
+    if discovery_info is None:
+        _LOGGER.error("This platform is only available through discovery")
+        return
+
+    compensation = discovery_info.get(CONF_COMPENSATION)
+    conf = hass.data[DATA_COMPENSATION][compensation]
+
+    async_add_entities(
+        [
+            CompensationSensor(
+                hass,
+                conf[CONF_ENTITY_ID],
+                conf.get(CONF_NAME),
+                conf.get(CONF_ATTRIBUTE),
+                conf[CONF_PRECISION],
+                conf[CONF_DEGREE],
+                conf[CONF_DATAPOINTS],
+                conf.get(CONF_UNIT_OF_MEASUREMENT),
+            )
+        ]
+    )
+
+
+class CompensationSensor(Entity):
+    """Representation of a Compensation sensor."""
+
+    def __init__(
+        self,
+        hass,
+        entity_id,
+        name,
+        attribute,
+        precision,
+        degree,
+        datapoints,
+        unit_of_measurement,
+    ):
+        """Initialize the Compensation sensor."""
+        self._entity_id = entity_id
+        self._name = name
+        self._precision = precision
+        self._attribute = attribute
+        self._unit_of_measurement = unit_of_measurement
+
+        self._points = []
+        for datapoint in datapoints:
+            match = re.match(MATCH_DATAPOINT, datapoint)
+            # we should always have x and y if the regex validation passed.
+            x_value, y_value = [float(v) for v in match.groups()]
+            self._points.append((x_value, y_value))
+
+        x_values, y_values = zip(*self._points)
+        self._coefficients = np.polyfit(x_values, y_values, degree)
```

Comment:
Should we find if users supply a rank deficient set of data?  For example, 

```
  - 0 -> 1
  - 0 -> 2
```

`numpy.polyfit` uses `warnings.warn` with a `RankWarning` in this case, and I couldn't find any place where homeassistant core handles warnings by default.  At the minimum I would recommend logging this warning using the level error or warning.  But in my opinion, the entity should not even be setup as it will be pure nonsense.  In the latter case, you would need to move this up to `async_setup_platform`.

Reply:
Yeah, I was thinking about that.  I was trying to find bad datasets to cause RankWarnings.  I have to make a few changes tonight anyways, I'll probably reorganize this.  I don't mind moving the poly generation into the setup.  It doesn't need to be where it is.

Reply:
Another wrinkle or several wrinkles actually, the example I gave you raises an exception and warns using `RunTimeWarning` because of a divide by zero. The issue is that the exact Exception raised depends on the backend linear algebra library in use.  On one system that I tested it raises `ValueError`, on another it raises `LinAlgError`  `LinAlgError` unfortunately does not inherit from `ValueError`.  I suppose you could catch both of those, and hope that catches all the possibilities?  it also does not warn using `RankWarning`, so this cannot be used as the overarching determinator.

In good news, if you use a different form, you can get it to consistently use `RankWarning`.

A summary:
```
import numpy as np
# raises at least LinAlgError and ValueError depending on linear algebra backend. Warns with RuntimeWarning
np.polyfit([0, 0], [1, 1], 1) 
# only warns with RankWarning
np.polyfit([1, 1], [1, 1], 1)

Reply:
If you're seeing this much variability across platforms it seems I may need to account for these errors when users report them.  For now, I'll add the try excepts for each of those errors/warnings.  I'm going to move the poly fits into the __init__.py and store the poly in data

Reply:
I think I've found a solution that should work.  If I set np.seterr(all='raise'), then all division warnings will come back as a FloatingPointError instead of a warning.  The np.linalg.LinAlgError is never raised in that instance.  This also keeps the rankwarning as a rank warning, which should do nothing other than raise the rank warning.

Reply:
I'll put a catch in for
```
except (np.linalg.LinAlgError, FloatingPointError, ValueError) as e:
```

and just post any warnings that don't fit in those errors and continue

Reply:
With `np.setrerr(all='raise')`, I only get `FloatingPointError` on both systems I tested, so I'm not sure you need the other two now :).  You should instead use the context manager https://numpy.org/doc/stable/reference/generated/numpy.errstate.html

The second example still only warns, so you will want to handle that case too.

Reply:
I'll remove the catches for ValueError and LinAlgError

Reply:
I've accounted for everything here